### PR TITLE
Lazy sessions

### PR DESCRIPTION
**Pull request by seebye**
Changes compared to the last pull request:

- restore tab if visible
- after restore:  forward the user to the 
  * next history entry if he came from a previous history entry
  * otherwise forward him to the previous history entry

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3345)

### CODE REVIEW

File: qutebrowser/browser/qutescheme.py [hunk 1]
```diff
@@ -425,6 +426,18 @@ def qute_settings(url):
...
 
+@add_handler('back')
+def qute_back(url):
+    """Handler for qute://back.
+
+    Simple page to free ram / lazy load a site, goes back on focusing the tab.
+    """
+    html = jinja.render(
+            'back.html',
+            title='Suspended: ' + urllib.parse.unquote(url.fragment()))
```
Comment (The-Compiler) [inline qutebrowser/browser/qutescheme.py]: Please only indent those by four spaces.

File: qutebrowser/html/back.html [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+{% block script %}
+const STATE_BACK = "back";
+const STATE_FORWARD = "forward";
+
+function switch_state(new_state) {
+	history.replaceState(
+		new_state,
+		document.title,
+		location.pathname+location.hash);
```
Comment (The-Compiler) [inline qutebrowser/html/back.html]: Please add spaces around the `+`.

File: qutebrowser/html/back.html [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+function go_back() {
+	switch_state(STATE_FORWARD);
+	history.back();
+}
```
Comment (The-Compiler) [inline qutebrowser/html/back.html]: Please add an empty line after this as well.

File: qutebrowser/html/back.html [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+		}
+		
+		document.addEventListener("visibilitychange", go_back);
+		break;
+}
+{% endblock %}
+
+{% block content %}
+<noscript><p>Javascript isn't enabled. So you need to manually go back in history to restore this tab.</p></noscript>
+<p>If you see this site something went wrong or you reached the end of the history or you disabled javascript.</p>
```
Comment (The-Compiler) [inline qutebrowser/html/back.html]: Please add commas before the "or"s.

File: qutebrowser/misc/sessions.py [hunk 5]
```diff
@@ -323,7 +331,18 @@ def save_last_window_session(self):
     def _load_tab(self, new_tab, data):
         """Load yaml data into a newly opened tab."""
         entries = []
-        for histentry in data['history']:
+        lazy_load = []
+        # use len(data['history'])
+        # -> dropwhile empty if not session.lazy_session
+        lazy_index = len(data['history'])
+        gen = itertools.chain(
+                itertools.takewhile(lambda _: not lazy_load,
```
Comment (The-Compiler) [inline qutebrowser/misc/sessions.py]: Please only indent those by four spaces.

**Comment by The-Compiler** [comment]
I just gave this a try again, but it doesn't seem to work properly - when I load a session, almost all tabs are unsuspended after a second or so.

**Comment by seebye** [comment]
I need more information, as it's working fine for me: 
https://webmshare.com/Ye0Yj

**Comment by seebye** [comment]
Okay, I guess you're using webkit aren't you?
(according to your comment it sounds like document.hidden is undefined
-> !document.hidden is always true)
I thought it's consider as legacy engine (as it's insecure as far as I know https://blogs.gnome.org/mcatanzaro/2016/02/01/on-webkit-security-updates/ ),
therefore I didn't focus on creating compatible code..
However in the latest commit I added code which checks whether "hidden" or "webkitHidden" is usable.
If neither hidden nor webkitHidden is usable the script will tell the user to update his engine.

**Comment by The-Compiler** [comment]
No, I'm using an up-to-date (Qt 5.9.3) QtWebEngine, equivalent to Chromium 56.

qutebrowser does not support old/legacy QtWebKit versions, and the [new QtWebKit](https://github.com/annulen/webkit/wiki) does support `document.hidden` just fine - so please revert your latest commit.

I don't know why you can't reproduce it, I'll try with the same steps you did in your video now.

**Comment by The-Compiler** [comment]
Can you please try loading this session file?

```
windows:
- active: true
  geometry: !!binary |
    AdnQywACAAAAAAAAAAAAGAAAAqoAAAGLAAAAAgAAABoAAAKoAAABiQAAAAAAAAAABVY=
  tabs:
  - history:
    - active: true
      original-url: https://bitbucket.org/portix/dwb/pull-requests/22/several-cleanups-to-increase-portability
      scroll-pos:
        x: 0
        y: 560
      title: ''
      url: https://bitbucket.org/portix/dwb/pull-requests/22/several-cleanups-to-increase-portability/diff
      zoom: 1.75
  - history:
    - active: true
      scroll-pos:
        x: 0
        y: 0
      title: qutebrowser | qutebrowser
      url: http://www.qutebrowser.org/
      zoom: 1.75
  - active: true
    history:
    - active: true
      scroll-pos:
        x: 0
        y: 6145
      title: New Features in Qt 5.5 - Qt Wiki
      url: http://wiki.qt.io/New_Features_in_Qt_5.5
      zoom: 1.75
```

for me, the first tab is suspended, but the second/third one aren't.

File: qutebrowser/html/back.html [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+		}
+		
+		document.addEventListener("visibilitychange", go_back);
+		break;
+}
+{% endblock %}
+
+{% block content %}
+<noscript><p>Javascript isn't enabled. So you need to manually go back in history to restore this tab.</p></noscript>
+<p>If you see this site something went wrong, or you reached the end of the history, or you disabled javascript.</p>
```
Comment (The-Compiler) [inline qutebrowser/html/back.html]: This text is kind of confusing while the suspended page is loading. Maybe something like this instead?

> Loading suspended page...
>
> If nothing happens, something went wrong or you disabled JavaScript.

**Comment by seebye** [comment]
> Can you please try loading this session file?

First and second tab suspended, third (focused) tab loaded.
So it works as expected at least on my notebook.

Tabs are instantly loaded aren't they?
I guess it's a timing issue / the javascript code is executed while the tab is still focused.
Could you try the latest version to confirm or to rule out my suspision?

**Comment by The-Compiler** [comment]
I can still reproduce it - looks like it's related to the `tabs.background` setting! When turning that off (which is the default), things seem fine again.

**Comment by The-Compiler** [comment]
Looking at this:

```diff
diff --git a/qutebrowser/html/back.html b/qutebrowser/html/back.html
index 46945ab65..a1ccb83b6 100644
--- a/qutebrowser/html/back.html
+++ b/qutebrowser/html/back.html
@@ -22,6 +22,7 @@ function go_forward() {
 }
 
 function prepare_restore() {
+  console.log(`prepare_restore of ${document.title}, hidden ${document.hidden}`);
        if (!document.hidden) {
                go_back();
                return;

```

It seems like `document.hidden` is `false` when the tab was never shown, which might be a QtWebEngine bug? This helps:

```diff
diff --git a/qutebrowser/mainwindow/tabbedbrowser.py b/qutebrowser/mainwindow/tabbedbrowser.py
index c78c0eced..e204e37fb 100644
--- a/qutebrowser/mainwindow/tabbedbrowser.py
+++ b/qutebrowser/mainwindow/tabbedbrowser.py
@@ -474,11 +474,9 @@ class TabbedBrowser(tabwidget.TabWidget):
         if background is None:
             background = config.val.tabs.background
         if background:
-            # Make sure the background tab has the correct initial size.
-            # With a foreground tab, it's going to be resized correctly by the
-            # layout anyways.
-            tab.resize(self.currentWidget().size())
-            self.tab_index_changed.emit(self.currentIndex(), self.count())
+            old_idx = self.currentIndex()
+            self.setCurrentWidget(tab)
+            self.setCurrentIndex(old_idx)
         else:
             self.setCurrentWidget(tab)
 
```

but I'm not sure what other side-effects it has.

**Comment by The-Compiler** [comment]
I opened #3366 for this now, as it seems like there's some deeper issue there.

Unfortunately, lazy sessions aren't really useful when they're sometimes not lazy... What do you think about checking when a `qute://back` tab was focused in Python instead, and going back there instead of doing it via JavaScript?

**Comment by seebye** [comment]
> looks like it's related to the tabs.background setting!
> It seems like document.hidden is false when the tab was never shown, which might be a QtWebEngine bug? This helps:
> `...`

In this case shouldn't the problem be solved if we replace
```python
new_tab = tabbed_browser.tabopen()
```
with
```python
new_tab = tabbed_browser.tabopen(background=False)
```
in SessionManager#load (qutebrowser/misc/sessions.py)?

The behavior of tabs.background shouldn't change according to it's description
> Open new tabs (middleclick/ctrl+click) in the background.

**Comment by The-Compiler** [comment]
Hmm, you're right, not sure why I didn't think of that 😆 

Thanks! I'd still like to have "real" lazy loading somewhen in the future, but until then, this is probably already a big improvement for people loading big sessions.

**Comment by edrex** [comment]
I just noticed this when I went into the qutebrowser tracker after a few "Open session, overload python, crash" cycles after a reboot. Don't want to spam, but super happy that this is in! Definitely fixes one of the big usability issues for me. Maybe at some point this should be default, or there could be a "safety valve" heuristic to lazy load if it looks like the number of tabs in the session are likely to cause a crash? Making things work smoothly in the default configuration is key IMO (don't get me wrong I love the customizability!)

**Comment by The-Compiler** [comment]
@edrex Why is this happening though? Are you running out of RAM?

**Comment by edrex** [comment]
@The-Compiler I commonly run upwards of 50 tabs in a session. In the steady state, it runs fine, especially with the various speed improvements around 1.0, but when they all try to start at once, sometimes all I see are blank white windows. I could script a stress test to reproduce the conditions for debugging. Are there existing startup stress tests in the suite?

**Comment by edrex** [comment]
I wonder if any of my settings are responsible, especially `tabs_are_windows`:

```
c.auto_save.session=True
c.statusbar.hide=True
c.completion.cmd_history_max_items = -1
c.tabs.show="multiple"
c.tabs.tabs_are_windows = True
```

**Comment by edrex** [comment]
Another thing that might be related is that sometimes I end up restoring the session into an i3 columns layout, which ends up squishing each window a bit with each additional window. Maybe all those resize events are causing a traffic jam.

**Comment by The-Compiler** [comment]
With `tabs_are_windows` you'll probably not really get lazy loading, as all your windows will be focused at some point. I don't think there are any benchmarks for opening new tabs (or loading sessions).

**Pull request merged**